### ISSUE TITLE
Add tutorial for UITextDisplayer

### ISSUE THREAD

**Issue by WINOFFRG** [shaka-project/shaka-player#4862]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?**
Yes

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
v4.3.2 with Custom UI

**Please ask your question**

I am trying to add subtitles container to my React application. I am using UITextDisplayer plugin for the same. I just created a new instance of that by passing my `videoElement` and `cuesContainer` element reference. However, I see the text no where on the DOM but instead on the video element. Is it some issue with my react application? Or the way I am using the plugin. I definitely see text being displayed on the video element but those span tags being missing in the DOM makes it difficult for me to access and style it.

![image](https://user-images.githubusercontent.com/61311751/210227545-d516b421-a81a-43fc-ae03-f4c36e27bba7.png)

**Comment by WINOFFRG** [comment]
Here is the code https://github.com/WINOFFRG/limetree/blob/d094363878472e8d42be37bf5b3a86266408c7b7/src/components/ControlsOverlay/layout.tsx#L35

Here is the preview: https://limetree-m9jfsbmm5-winoffrg.vercel.app/

Please set the time to t=150 `$('video').currentTime = 150;`

**Comment by WINOFFRG** [comment]
@avelad Can you please take a look at this one? I think its issue with how I am setting up captions on my custom UI. Currently, I guess native cue methods are being used as they are being displayed on the video itself but not anywhere in the DOM.

**Comment by joeyparrish** [comment]
The player isn't using the UI displayer plugin.  I see cues in the video element's shadow root instead.

**Comment by joeyparrish** [comment]
Looking at your code (https://github.com/WINOFFRG/limetree/blob/d094363878472e8d42be37bf5b3a86266408c7b7/src/components/ControlsOverlay/layout.tsx#L35), you aren't configuring the text displayer in the player.  Creating the instance isn't enough.

Try this:

```js
shakaPlayer.configure('textDisplayFactory', () => new shaka.text.UITextDisplayer(
    video, cuesContainer.current));
```

The player instance will create it and destroy it as-needed, so you can drop the callback at the end of useEffect():

```js
        return () => {
            if (textCuesContainer) textCuesContainer.destroy();
        };
```

Does this help?

**Comment by WINOFFRG** [comment]
Ahhhh, Thanks a lot. Literally banged my head all day on this thinking that this might be issue related to react mutations in virtual DOM. BTW, Am I missing some docs here? Steps to configure or maybe some extensions are they documented anywhere?

**Comment by joeyparrish** [comment]
It's in the API docs, but there's nothing in the tutorials to explain it in prose.

https://shaka-player-demo.appspot.com/docs/api/shaka.extern.html#.PlayerConfiguration
https://shaka-player-demo.appspot.com/docs/api/shaka.extern.TextDisplayer.html#.Factory

Would you be interested in adding a short tutorial on the use of UITextDisplayer?  I would be happy to review a PR for that.

Any new tutorial should be added to the index for a link to show up in the generated docs: https://github.com/shaka-project/shaka-player/blob/main/docs/tutorials/index.json

**Comment by WINOFFRG** [comment]
I almost missed the configuration steps, Thanks. 

> Would you be interested in adding a short tutorial on the use of UITextDisplayer? I would be happy to review a PR for that.

Sure, please assign it to me.

**Comment by joeyparrish** [comment]
Done!

**Comment by loicraux** [comment]
It would be nice to explain in the tutorial how you can subclass `UITextDisplayer` to override the cues styles...
My player is trying to apply a `background-color` to the subtitles (webvtt or ttml), I thought of reimplementing the `append` method of `UITextDisplayer`, but not sure when this method is called, if it takes effect immediately or not, etc ...

So I ended up using CSS, something like this:
``css
#player_element_Id .videoContainer .shaka-text-container span:not([class]) {
    background-color: black !important;
}
```

It works, but it's very fragile (I try to avoid spans that are not cues)... I would have preferred to make changes like `cue.backgroundColor = 'black'` in the code...

**Comment by loicraux** [comment]
> IMO it is better to override CSS than to override methods, when possible.  If we need to add classes to those spans, that seems like a good way to allow more targeted CSS.

Yes I thought about that too. I'll file an issue and PR this week.

**Comment by YamaGroupTech** [comment]
Hello guys,

I'm facing a problem using ReactShakaPLayer with this import `@mkhuda/react-shaka-player` when loading the subtitles sometimes it creates me the class .shaka-text-container and sometimes it doesn't. 

Basically, when the DOM is loaded I load the file and then I set the params to have cues. Then I apply css to the class. 

Maybe is due to a rendering issue ?



```
    useEffect(() => {
        async function initVideo(){
            
            player.configure('streaming.alwaysStreamText', true)

             let playerContainer = document.getElementById('video-container');

            player.configure('textDisplayFactory', () => new shaka.text.UITextDisplayer(
                videoElement,
                 playerContainer
            ));


            player.setVideoContainer(playerContainer);


            await player.load(file)


            const subtitleUri = "url";
            player.addTextTrackAsync(subtitleUri, 'en', 'subtitles', 'text/vtt');
            
            player.selectTextTrack(player.getTextTracks()[0])

            player.setTextTrackVisibility(true);

        }

        if(loadedDom && data.file && player ) {

            shaka.polyfill.installAll();
            if (shaka.Player.isBrowserSupported()) {
                initVideo()
            } else {
                console.error('Browser not supported!');
            }

        }
    }, [loadedDom, player, data])
```


My return 


```
<div id='video-container'>
                        <ReactShakaPlayer
                            // webkit-playsinline="true"
                            // playsInline
                            // player={player}
                            // videoElement={videoElement}

                            id="videoLoop"
                            playerClassName="shaka-player-1"
                            onLoad={(player) => setVideoPlayer(player)}
                            onEnded={() => endedVideo()}
                            // uiConfig={config}
                            muted
                            loop={loop}
                        />
                    </div>
```


I tried a standalone using just html and js and it seems to work just fine by passing those configs.

**Comment by BruceZunino** [comment]
Hello guys,

I'm facing a problem using ReactShakaPLayer with this import @mkhuda/react-shaka-player when loading the subtitles sometimes it creates me the class .shaka-text-container and sometimes it doesn't.

Basically, when the DOM is loaded I load the file and then I set the params to have cues. Then I apply css to the class.

Maybe is due to a rendering issue ?

```
    useEffect(() => {
        async function initVideo(){
            
            player.configure('streaming.alwaysStreamText', true)

             let playerContainer = document.getElementById('video-container');

            player.configure('textDisplayFactory', () => new shaka.text.UITextDisplayer(
                videoElement,
                 playerContainer
            ));


            player.setVideoContainer(playerContainer);


            await player.load(file)


            const subtitleUri = "url";
            player.addTextTrackAsync(subtitleUri, 'en', 'subtitles', 'text/vtt');
            
            player.selectTextTrack(player.getTextTracks()[0])

            player.setTextTrackVisibility(true);

        }

        if(loadedDom && data.file && player ) {

            shaka.polyfill.installAll();
            if (shaka.Player.isBrowserSupported()) {
                initVideo()
            } else {
                console.error('Browser not supported!');
            }

        }
    }, [loadedDom, player, data])
```
My return

```
<div id='video-container'>
                        <ReactShakaPlayer
                            // webkit-playsinline="true"
                            // playsInline
                            // player={player}
                            // videoElement={videoElement}

                            id="videoLoop"
                            playerClassName="shaka-player-1"
                            onLoad={(player) => setVideoPlayer(player)}
                            onEnded={() => endedVideo()}
                            // uiConfig={config}
                            muted
                            loop={loop}
                        />
                    </div>
```
I tried a standalone using just html and js and it seems to work just fine by passing those configs.

**Comment by WINOFFRG** [comment]
Hi @BruceZunino ! Recently I have been facing similar issue and created it on [VideoDev](https://video-dev.slack.com/archives/C01QRAFHLQK/p1690776736808769) Slack,  another member on the slack has also reported similar issue. In my case the captions are visible on native video element and not on the UI. I have tried preious major versions as well till 4.0.0 there also it didn't seem to work for me but does work in vanillajs and demo UI. I feel it's some issue with the way we are configuring it.

Your configuration looks fine to me and here's mine similar to yours.
 https://github.com/WINOFFRG/limeplay/blob/d04209423dd69c70f64dff62a06e71fbf94e2f52/home/src/components/Player/PlayerOverlay/index.tsx#L98-L105
 
Can be viewed here https://limeplay.me/ You would see the captions are appearing on the video element, resolving back to this same github issue I created. Once fixed, I think I will add the tutorial for it on priority.

**Comment by BruceZunino** [comment]
Hi @WINOFFRG and thanks for your answer.

I ended up deleting the component and doing all the config by using the basic js example from Shaka.
I followed this example: https://shaka-player-demo.appspot.com/docs/api/tutorial-basic-usage.html

I believe now I do not have any complication, but I really wish that we could be able to use the React npm because it would be much easier.

Anyways, I hope you guys find this helpful and if something changes in the near future, I would consider it.
